features/FEATURE_LWIP/lwip-interface/lwip_stack.c

```diff
@@ -1123,6 +1125,53 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
             }
             return 0;
 
+        case NSAPI_ADD_MEMBERSHIP:
+        case NSAPI_DROP_MEMBERSHIP: {
+            if (optlen != sizeof(nsapi_ip_mreq_t)) {
+                return NSAPI_ERROR_UNSUPPORTED;
+            }
+            err_t igmp_err;
+            nsapi_ip_mreq_t *imr = optval;
+
+            /* Check interface address type matches group, or is unspecified */
+            if (imr->imr_interface.version != NSAPI_UNSPEC && imr->imr_interface.version != imr->imr_multiaddr.version) {
+                return NSAPI_ERROR_PARAMETER;
+            }
+
+            ip_addr_t if_addr;
+            ip_addr_t multi_addr;
+
+            /* Convert the group address */
+            if (!convert_mbed_addr_to_lwip(&multi_addr, &imr->imr_multiaddr)) {
+                return NSAPI_ERROR_PARAMETER;
+            }
+
+            /* Convert the interface address, or make sure it's the correct sort of "any" */
+            if (imr->imr_interface.version != NSAPI_UNSPEC) {
+                if (!convert_mbed_addr_to_lwip(&if_addr, &imr->imr_interface)) {
+                    return NSAPI_ERROR_PARAMETER;
+                }
+            } else {
+                ip_addr_set_any(IP_IS_V6(&if_addr), &if_addr);
+            }
+
+            igmp_err = NSAPI_ERROR_UNSUPPORTED;
+            #if LWIP_IPV4
+            if (IP_IS_V4(&if_addr)) {
+                igmp_err = (NSAPI_ADD_MEMBERSHIP ?  igmp_joingroup : igmp_leavegroup)
+                                        (ip_2_ip4(&if_addr), ip_2_ip4(&multi_addr));
+            }
+            #endif
+            #if LWIP_IPV6
+            if (IP_IS_V6(&if_addr)) {
+                igmp_err = (NSAPI_ADD_MEMBERSHIP ? mld6_joingroup : mld6_leavegroup)
+                                        (ip_2_ip6(&if_addr), ip_2_ip6(&multi_addr));
+            }
+            #endif
+
+            return igmp_err;
```

Comment:
Need to do error conversion from lwip to NSAPI here. I guess you could then initialise igmp_err to ERR_VAL to get NSAPI_ERROR_PARAMETER - no err_t currently maps to NSAPI_ERROR_UNSUPPORTED.

Seems reasonable - I kind of like distinguishing "I don't know this option" from "I don't like what you specified for this option". I'd argue all existing size checks returning UNSUPOPRTED should be PARAMETER too.